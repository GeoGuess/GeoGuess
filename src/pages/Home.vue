<template>
    <ContentPage class="home-page">
        <section class="home-page__main">
            <v-container class="home-page__main__container" fluid>
                <v-layout class="home-page__main__layout">
                    <div class="home-page__traveler-container">
                        <img
                            class="home-page__traveler-img"
                            src="../assets/home/traveller.svg"
                        />
                    </div>
                    <v-layout class="home-page__main__content">
                        <div class="box">
                            <SearchBox :dialogCustomOpen="dialogCustomOpen" />
                        </div>
                    </v-layout>
                </v-layout>
            </v-container>
            <v-btn id="btnMaps" href="#maps-container" large fab color="secondary" dark>
                <v-icon>mdi-arrow-down</v-icon>
            </v-btn>
        </section>
        <MapsContainer />
    </ContentPage>
</template>

<script>
import SearchBox from '@/components/home/SearchBox';
import ContentPage from '@/components/page/ContentPage';
import { GAME_MODE } from '../constants';
import MapsContainer from '@/components/home/MapsContainer.vue';
export default {
    components: {
        ContentPage,
        SearchBox,
        MapsContainer,
    },
    props: {
        dialogCustomOpen: Boolean,
    },
    mounted() {
        if (this.$route.params && this.$route.params.partyParams) {
            const params = atob(this.$route.params.partyParams)
                .split(',')
                .map((val) => parseFloat(val));

            if (params.length >= 12 && params.length % 2 === 0) {
                const difficulty = params[0];
                const timeLimitation = params[1];
                const rounds = new Array((params.length - 2) / 2)
                    .fill(0)
                    .map((_, round) => {
                        const index = (round + 1) * 2;
                        return params.slice(index, index + 2);
                    });

                this.$router.push({
                    name: 'street-view',
                    params: {
                        modeSelected: GAME_MODE.CLASSIC,
                        time: timeLimitation,
                        difficulty: difficulty,
                        roundsPredefined: rounds,
                    },
                });
            }
        }
    },
};
</script>

<style scoped lang="scss">
.v-application--is-rtl .home-page__traveler-img{
    transform: scaleX(-1)

}
.home-page {
    .demo-alert {
        position: absolute;
        z-index: 1;
        width: 100%;
    }
    background-color: var(--v-home-base);
    .home-page__main {
        position: relative;
      .theme--light & .home-page__main__container {
        background: url('../assets/home/world.svg');
      }
      .theme--dark & .home-page__main__container {
        background: url('../assets/home/world-dark.svg');
      }
        .home-page__main__container {
            font-size: 1.2rem;
            padding: 0;
            margin: 0;
            width: 100%;
            background-size: cover;
            background-position: top;
            .home-page__main__layout {
                height: calc(70vh - 100px);
                flex-wrap: nowrap;
                justify-items: end;
                .box {
                    margin: auto;
                    width: 35vw;
                    max-width: 100%;
                }

                .home-page__main__content {
                    min-width: 65%;
                }
                .home-page__traveler-container {
                    height: auto;
                    width: 100%;
                    max-width: 50vw;
                    display: flex;
                    justify-content: flex-start;
                    .home-page__traveler-img {
                        max-width: 30vw;
                        max-height: 60vh;
                    }
                }
            }
        }
        #btnMaps {
            position: absolute;
            margin: auto;
            bottom: 0.4rem;
            left: 0;
            right: 0;
        }
    }
    
}
@media (max-width: 1300px) and (min-width: 600px) {
    .home-page
        .home-page__main
        .home-page__main__container
        .home-page__main__layout
        .box {
        width: 60vw;
    }
}
@media (max-width: 660px) {
    .home-page {
        background-color: #ded3af;
        .home-page__main .home-page__main__container {
            .home-page__main__layout {
                flex-direction: column-reverse;

                .box {
                    width: 90vw;
                }
            }
        }
    }
}


@media (max-height: 550px) {
    .home-page
        .home-page__main
        .home-page__main__container
        .home-page__main__layout
        .box {
        margin: 10vh;
    }

    #btnMaps {
        display: none;
    }
}
</style>
