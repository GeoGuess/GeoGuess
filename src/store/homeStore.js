import axios from '@/plugins/axios';
import { validURL } from '@/utils';
import { getLocateString, isGeoJSONValid } from '../utils';
import * as MutationTypes from './mutation-types';
import i18n from '../lang';
import IndexedDBService from '../plugins/IndexedDBService';
import { GeoMap, GeoMapCustom } from '../models/GeoMap';

export default {
    state: () => ({
        map: new GeoMapCustom(),
        loadingGeoJson: false,
        errorMessage: null,
        listMaps: [],
        listAreas: [],
        customsMaps: [],
        history: [],
        streamerMode: localStorage.getItem('streamerMode') === 'true',
    }),
    mutations: {
        [MutationTypes.HOME_SET_GEOJSON](state, geojson) {
            const map = new GeoMapCustom();
            map.name = state.map.name;
            map.geojson = geojson;
            state.map = map;
        },
        [MutationTypes.HOME_SET_NAME_GEOJSON](state, name) {
            state.map.name = name;
        },
        [MutationTypes.HOME_SET_MAP](state, map) {
            state.map = map;
        },
        [MutationTypes.HOME_SET_LISTS](state, lists) {
            state.listMaps = lists.maps;
            state.listAreas = lists.areas;
        },
        [MutationTypes.HOME_SET_LISTS_CUSTOMMAPS](state, customsMaps) {
            state.customsMaps = customsMaps;
        },
        [MutationTypes.HOME_SET_HISTORY](state, history) {
            state.history = history;
        },
        [MutationTypes.HOME_SET_STREAMER_MODE](state, streamerMode) {
            state.streamerMode = streamerMode;
            localStorage.setItem('streamerMode', streamerMode);
        },

        [MutationTypes.HOME_SET_STATUS_GEOJSON](state, status) {
            state.loadingGeoJson = status;
        },
        [MutationTypes.HOME_SET_GEOJSON_ERROR](state, error) {
            state.errorMessage = error;
        },
    },

    getters: {
        geoJsonString(state) {
           if (!state.map || !state.map.geojson) {
                return '';
            }
            return JSON.stringify(state.map.geojson, null, 2);
        },
        geoJson(state) {
            return state.map.geojson;
        },
        isValidGeoJson(state) {
           if (!state.map || !state.map.geojson) {
                return null;
            }
            return isGeoJSONValid(state.map.geojson);
        },
        maps(state) {
            return state.customsMaps
                .map((map) => Object.assign(new GeoMapCustom(), map))
                .concat(
                    state.listMaps.map((map) =>
                        Object.assign(new GeoMap(), map)
                    )
                );
        },
        areasList(state) {
            return state.listAreas.map((map) => ({
                ...map,
                nameLocate: getLocateString(map, 'name', i18n.locale),
                descriptionLocate: getLocateString(
                    map,
                    'description',
                    i18n.locale
                ),
            }));
        },
        nbPlaceVisits(state) {
            return state.history.reduce(
                (a, { rounds }) => a + rounds.length,
                0
            );
        },
    },

    actions: {
        setStreamerMode({ dispatch, commit }, value) {
            commit(MutationTypes.HOME_SET_STREAMER_MODE, value);
            dispatch(
                'alertStore/setAlert',
                value && {
                    title: 'Home.streamerModeActivate',
                    subtitle: 'Home.streamerModeDetails',
                    color: 'streamerMode',
                    icon: 'mdi-twitch',
                },
                { root: true }
            );
        },
        loadPlaceGeoJSON({ commit, state }, place) {
            if (place != null && place != '') {
                if (state.loadingGeoJson) {
                    return;
                }
                commit(MutationTypes.HOME_SET_STATUS_GEOJSON, true);

                commit(MutationTypes.HOME_SET_GEOJSON, null);

                axios
                    .get(
                        `https://nominatim.openstreetmap.org/search/${encodeURIComponent(
                            place.toLowerCase()
                        )}?format=geojson&limit=1&polygon_geojson=1`
                    )
                    .then((res) => {
                        if (
                            res &&
                            res.status === 200 &&
                            res.data.features.length > 0
                        ) {
                            let feature = res.data.features[0];
                            commit(MutationTypes.HOME_SET_GEOJSON, feature);
                            return;
                        }
                        commit(
                            MutationTypes.HOME_SET_GEOJSON_ERROR,
                            'No Found Location'
                        );
                    })
                    .finally(() => {
                        commit(MutationTypes.HOME_SET_STATUS_GEOJSON, false);
                    });
            }
        },
        async loadGeoJsonFromUrl({ commit }, url) {
            if (validURL(url)) {
                // if gist url get raw
                if (RegExp('^(https?://)?gist.github.com/').test(url)) {
                    let urlSplit = url.split('/');
                    if (
                        urlSplit.length > 3 &&
                        urlSplit[urlSplit.length - 1] !== 'raw'
                    ) {
                        urlSplit[urlSplit.length - 3] =
                            'gist.githubusercontent.com';
                        urlSplit.push('raw');
                        url = urlSplit.join('/');
                    }
                }
                const geojson = await axios
                    .get(url)
                    .then((res) => {
                        if (res.status === 200 && res.data) {
                            if (typeof res.data === 'object') {
                                return res.data;
                            } else {
                                return JSON.parse(res.data);
                            }
                        }
                    })
                    .catch((err) => {
                        // eslint-disable-next-line no-console
                        console.log(err);
                    });

                commit(MutationTypes.HOME_SET_GEOJSON, geojson);
            }
        },
        setMapLoaded({ commit }, map) {
            commit(MutationTypes.HOME_SET_MAP, map);
        },
        setGeoJson({ commit }, geojson) {
            commit(MutationTypes.HOME_SET_GEOJSON, geojson);
        },
        async saveGeoJson({ state, dispatch }) {
            await state.map.save();
            dispatch('getListMapsCustoms');
            dispatch(
                'alertStore/setAlert',
                {
                    title: 'Home.mapSavedAlert.title',
                    subtitle: 'Home.mapSavedAlert.subtitle',
                },
                { root: true }
            );
        },
        setGeoJsonString({ commit }, geojson) {
            let obj = null;
            if (geojson !== '') {
                obj = JSON.parse(geojson);
            }
            commit(MutationTypes.HOME_SET_GEOJSON, obj);
        },
        async getListMaps({ commit }) {
            const data = await axios
                .get(
                    process.env.VUE_APP_LIST_MAPS_JSON_URL ||
                        'https://maps.geoguess.games/maps.json',
                    {
                        cache: {
                            maxAge: 1000,
                        },
                    }
                )
                .then((res) => res.data);

            commit(MutationTypes.HOME_SET_LISTS, data);
        },
        async getListMapsCustoms({ commit }) {
            const customsMap = await Promise.resolve(
                IndexedDBService.loadDb().then(async () => {
                    return await IndexedDBService.getAllMaps();
                })
            );
            commit(MutationTypes.HOME_SET_LISTS_CUSTOMMAPS, customsMap);
        },
        loadHistory({ commit }) {
            const history = localStorage.getItem('history')
                ? JSON.parse(localStorage.getItem('history'))
                : [];
            commit(MutationTypes.HOME_SET_HISTORY, history);
        },
    },
};
