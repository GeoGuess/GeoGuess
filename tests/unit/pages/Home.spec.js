import Home from '@/pages/Home.vue';
import { createLocalVue, shallowMount } from '@vue/test-utils';
import appInit from '../testutils/appInit';
import Vuex from 'vuex';
import homeStore from '../../../src/store/homeStore';

const args = appInit(createLocalVue(), false);
const $router = {
    push: jest.fn(),
};

describe('Home.vue', () => {
    let store, actions;
    beforeEach(() => {
        actions = {
            getListMaps: jest.fn(),
            getListMapsCustoms: jest.fn(),
        };
        store = new Vuex.Store({
            modules: {
                homeStore: {
                    state: homeStore.state,
                    getters: homeStore.getters,
                    actions,
                },
            },
        });
    });
    it('test mounted', () => {
        const wrapper = shallowMount(Home, {
            ...args,
            store,
            mocks: {
                $route: {
                    params: {},
                },
                $router,
            },
        });
        expect($router.push).not.toBeCalled();

        expect(actions.getListMaps).toBeCalled();
        expect(actions.getListMapsCustoms).toBeCalled();

        expect(wrapper).toMatchSnapshot();
    });
    it('test mounted', () => {
        const partyParams =
            'MjAwMCwzMDAsMzkuOTgyOTgxMzM0MTE0MDIsMjMuNjE4NDQ1MjIxOTg0OTgsNzAuMDAxNDM5OTEwOTEwNDUsMTkuNTYwMjI5NzM3Njk5MTcsNTguNDEyNzQ4MTQ2ODQ3MzIsNi45MTI3OTk3MzYzODg2NjYsNjIuMzQxODY2Njg2NDkxNTQsLTYuMjYxNDk1MTQyNDk0MjI2LDY3LjA5OTY2MzkwODQ4Mzc2LDE5LjUxMjI4NDk3NTUwODY0';
        const wrapper = shallowMount(Home, {
            ...args,
            store,
            mocks: {
                $route: {
                    params: {
                        partyParams,
                    },
                },
                $router,
            },
        });
        expect($router.push).toHaveBeenCalledWith({
            name: 'street-view',
            params: {
                time: 300,
                modeSelected: 'classic',
                difficulty: 2000,
                roundsPredefined: [
                    [39.98298133411402, 23.61844522198498],
                    [70.00143991091045, 19.56022973769917],
                    [58.41274814684732, 6.912799736388666],
                    [62.34186668649154, -6.261495142494226],
                    [67.09966390848376, 19.51228497550864],
                ],
            },
        });

        expect(actions.getListMaps).toBeCalled();
        expect(actions.getListMapsCustoms).toBeCalled();

        expect(wrapper).toMatchSnapshot();
    });
});
