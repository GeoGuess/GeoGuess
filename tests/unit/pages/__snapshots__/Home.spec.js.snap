// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home.vue test mounted 1`] = `
<contentpage-stub class="home-page">
  <section class="home-page__main">
    <v-container-stub tag="div" fluid="true" class="home-page__main__container">
      <v-layout-stub tag="div" class="home-page__main__layout">
        <div class="home-page__traveler-container"><img src="../assets/home/traveller.svg" class="home-page__traveler-img"></div>
        <v-layout-stub tag="div" class="home-page__main__content">
          <div class="box">
            <searchbox-stub></searchbox-stub>
          </div>
        </v-layout-stub>
      </v-layout-stub>
    </v-container-stub>
    <v-btn-stub color="secondary" dark="true" tag="button" activeclass="" href="#areas" large="true" fab="true" type="button" id="btnMaps">
      <v-icon-stub>mdi-arrow-down</v-icon-stub>
    </v-btn-stub>
  </section>
  <v-container-stub id="areas" tag="div">
    <h2>Areas</h2>
  </v-container-stub>
  <section class="sliders-container">
    <v-slide-group-stub valuecomparator="function deepEqual(a, b) {
          if (a === b) return true;

          if (a instanceof Date &amp;&amp; b instanceof Date &amp;&amp; a.getTime() !== b.getTime()) {
            // If the values are Date, compare them as timestamps
            return false;
          }

          if (a !== Object(a) || b !== Object(b)) {
            // If the values aren't objects, they were already checked for equality
            return false;
          }

          var props = Object.keys(a);

          if (props.length !== Object.keys(b).length) {
            // Different number of props, don't bother to check
            return false;
          }

          return props.every(function (p) {
            return deepEqual(a[p], b[p]);
          });
        }" activeclass="v-slide-item--active" tag="div" mobilebreakpoint="1264" nexticon="$next" previcon="$prev" showarrows="always"></v-slide-group-stub>
  </section>
  <v-container-stub tag="div">
    <h2>Maps</h2>
  </v-container-stub>
  <section id="maps"></section>
</contentpage-stub>
`;

exports[`Home.vue test mounted 2`] = `
<contentpage-stub class="home-page">
  <section class="home-page__main">
    <v-container-stub tag="div" fluid="true" class="home-page__main__container">
      <v-layout-stub tag="div" class="home-page__main__layout">
        <div class="home-page__traveler-container"><img src="../assets/home/traveller.svg" class="home-page__traveler-img"></div>
        <v-layout-stub tag="div" class="home-page__main__content">
          <div class="box">
            <searchbox-stub></searchbox-stub>
          </div>
        </v-layout-stub>
      </v-layout-stub>
    </v-container-stub>
    <v-btn-stub color="secondary" dark="true" tag="button" activeclass="" href="#areas" large="true" fab="true" type="button" id="btnMaps">
      <v-icon-stub>mdi-arrow-down</v-icon-stub>
    </v-btn-stub>
  </section>
  <v-container-stub id="areas" tag="div">
    <h2>Areas</h2>
  </v-container-stub>
  <section class="sliders-container">
    <v-slide-group-stub valuecomparator="function deepEqual(a, b) {
          if (a === b) return true;

          if (a instanceof Date &amp;&amp; b instanceof Date &amp;&amp; a.getTime() !== b.getTime()) {
            // If the values are Date, compare them as timestamps
            return false;
          }

          if (a !== Object(a) || b !== Object(b)) {
            // If the values aren't objects, they were already checked for equality
            return false;
          }

          var props = Object.keys(a);

          if (props.length !== Object.keys(b).length) {
            // Different number of props, don't bother to check
            return false;
          }

          return props.every(function (p) {
            return deepEqual(a[p], b[p]);
          });
        }" activeclass="v-slide-item--active" tag="div" mobilebreakpoint="1264" nexticon="$next" previcon="$prev" showarrows="always"></v-slide-group-stub>
  </section>
  <v-container-stub tag="div">
    <h2>Maps</h2>
  </v-container-stub>
  <section id="maps"></section>
</contentpage-stub>
`;
